library/libwaku.nim

```diff
@@ -54,47 +68,80 @@ proc relayEventCallback(ctx: ptr Context): WakuRelayHandler =
 
     try:
       let event = $JsonMessageEvent.new(pubsubTopic, msg)
-      cast[WakuCallBack](ctx[].eventCallback)(
-        RET_OK, unsafeAddr event[0], cast[csize_t](len(event)), ctx[].eventUserData
-      )
+      foreignThreadGc:
+        cast[WakuCallBack](ctx[].eventCallback)(
+          RET_OK, unsafeAddr event[0], cast[csize_t](len(event)), ctx[].eventUserData
+        )
     except Exception, CatchableError:
       let msg = "Exception when calling 'eventCallBack': " & getCurrentExceptionMsg()
-      cast[WakuCallBack](ctx[].eventCallback)(
-        RET_ERR, unsafeAddr msg[0], cast[csize_t](len(msg)), ctx[].eventUserData
-      )
+      foreignThreadGc:
+        cast[WakuCallBack](ctx[].eventCallback)(
+          RET_ERR, unsafeAddr msg[0], cast[csize_t](len(msg)), ctx[].eventUserData
+        )
 
 ### End of not-exported components
 ################################################################################
 
+################################################################################
+### Library setup
+
+# Every Nim library must have this function called - the name is derived from
+# the `--nimMainPrefix` command line option
+proc NimMain() {.importc.}
+
+# To control when the library has been initialized
+var initialized: Atomic[bool]
+
+if defined(android):
+  # Redirect chronicles to Android System logs
+  when compiles(defaultChroniclesStream.outputs[0].writer):
+    defaultChroniclesStream.outputs[0].writer =
+      proc (logLevel: LogLevel, msg: LogOutputStr) {.raises: [].} =
+        echo logLevel, msg
+
+
+### End of library setup
+################################################################################
+
 ################################################################################
 ### Exported procs
+proc waku_setup() {.dynlib, exportc.} =
+  NimMain()
+  if not initialized.load:
+    initialized.store(true)
+
+    # TODO: ask Ivan what is nimGC_setStackBottom for
+    when declared(nimGC_setStackBottom):
+      var locals {.volatile, noinit.}: pointer
+      locals = addr(locals)
+      nimGC_setStackBottom(locals)
```

Comment:
I had to introduce this function `waku_setup` in the bindings.

When I tried to use the original `waku_init` function, Android crashed. Looks like NimMain() (at least on android) needs to be called before any nim logic is executed, because the code `if not initialized.exchange(true):` that existed before did fail. I imagine it might be related to the `initialized` atomic being setup by `NimMain()`? 

Anyway, I think I heard/read somewhere that `NimMain()` function is not required in Nim 2.0? so perhaps on that version we can get rid of the existence of `waku_setup` then.

Also, I couldn't find docs for `nimGC_setStackBottom`. Is it really necessary?